src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using ICSharpCode.Decompiler.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CodeCleanup;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServerIndexFormat;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.RemoveUnnecessaryImports;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.ExternalAccess.Razor;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(RazorComponentInfoHandler)), Shared]
+[Method(RazorComponentInfoName)]
+internal sealed class RazorComponentInfoHandler : ILspServiceRequestHandler<RazorComponentInfoParams, RazorComponentInfo?>
+{
+    public const string RazorComponentInfoName = "roslyn/razorComponentInfo";
+    private readonly IGlobalOptionService _globalOptions;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public RazorComponentInfoHandler(IGlobalOptionService globalOptions)
+    {
+        _globalOptions = globalOptions;
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public async Task<RazorComponentInfo?> HandleRequestAsync(RazorComponentInfoParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        var project = context.Solution?.GetProject(request.Project);
+
+        if (project is null)
+        {
+            return null;
+        }
+
+        var document = context.Solution.GetDocument(request.Document);
+        var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        var syntaxTree = semanticModel.SyntaxTree;
+        var root = (CompilationUnitSyntax)syntaxTree.GetRoot(cancellationToken);
+
+        var (methods, fields) = GetInfoInsideRazorDocument(root, semanticModel, cancellationToken);
```

Comment:
It would be good to have a comment somewhere about what this endpoint actually does and is used for? I don't know why Razor wants to know about all of the method calls and fields used in a C# file.